File: qqrobot-spring-boot-starter/src/main/java/de/honoka/qqrobot/starter/component/session/RobotSession.java
package de.honoka.qqrobot.starter.component.session;

import de.honoka.qqrobot.framework.api.model.RobotMultipartMessage;
import de.honoka.sdk.util.code.ThrowsConsumer;
import lombok.Getter;
import lombok.Setter;
import lombok.experimental.Accessors;

import java.util.concurrent.TimeUnit;

/**
 * 会话类，用于记录处于会话状态的qq号和群号，以及这些号码在进入会话状态以后发送的信息
 */
@Getter
@Setter
@Accessors(chain = true)
public class RobotSession {

    private Long group;

    private long qq;

    private volatile RobotMultipartMessage reply;

    /**
     * 会话过程中要执行的操作
     */
    private ThrowsConsumer<RobotSession> action;

    /**
     * 超时操作
     */
    private ThrowsConsumer<RobotSession> onTimeout;

    /**
     * 所属的会话管理器
     */
    private final SessionManager sessionManager;

    public static class TimeoutException extends Exception {}

    //package-private
    RobotSession(Long group, long qq, SessionManager sessionManager) {
        this.qq = qq;
        this.group = group;
        this.sessionManager = sessionManager;
    }

    /**
     * 等待某个会话的回复，超时抛出超时异常
     * @param timeout 超时时间，单位为秒
     * @return  回复
     */
    public RobotMultipartMessage waitingForReply(int timeout)
            throws TimeoutException {
        reply = null;  //等待回复前，先忽略已有的回复
        int i = 0;  //已等待秒数
        while(reply == null) {
            try {
                TimeUnit.SECONDS.sleep(1);
                i++;
                if(i >= timeout) throw new TimeoutException();
            } catch(Exception e) {
                throw new TimeoutException();
            }
        }
        return reply;
    }

    public void reply(RobotMultipartMessage message) {
        sessionManager.getFramework().reply(group, qq, message);
    }

    public void reply(String message) {
        sessionManager.getFramework().reply(group, qq, message);
    }

    //package-private
    public void run() {
        sessionManager.openSession(this);
        try {
            action.accept(this);
        } catch(Throwable t) {
            //noinspection ConstantConditions
            if(!(t instanceof TimeoutException)) throw t;
            if(onTimeout != null) {
                onTimeout.accept(this);
            } else {
                sessionManager.getFramework().reply(group, qq,
                        "会话已超时关闭");
            }
        } finally {
            close();
        }
    }

    public void close() {
        sessionManager.closeSession(this);
    }
}

File: qqrobot-spring-boot-starter/src/main/java/de/honoka/qqrobot/starter/component/session/RobotSession.kt
package de.honoka.qqrobot.starter.component.session

import de.honoka.qqrobot.framework.api.model.RobotMultipartMessage
import java.io.Closeable
import java.util.concurrent.TimeUnit

/**
 * 会话类，用于记录处于会话状态的qq号和群号，以及这些号码在进入会话状态以后发送的信息
 */
class RobotSession private constructor(
    val group: Long?,
    val qq: Long,
    /**
     * 所属的会话管理器
     */
    private val sessionManager: SessionManager
) : Closeable {
    
    companion object {
        
        @JvmSynthetic
        internal fun of(group: Long?, qq: Long, sessionManager: SessionManager): RobotSession = run {
            RobotSession(group, qq, sessionManager)
        }
    }
    
    data class Action(
        
        /**
         * 会话过程中要执行的操作
         */
        var action: (RobotSession.() -> Unit)? = null,
        
        /**
         * 超时操作
         */
        var onTimeout: (RobotSession.() -> Unit)? = null
    )
    
    class TimeoutException : Exception()
    
    @set:JvmName("setReply")
    @get:JvmSynthetic
    @Volatile
    internal var reply: RobotMultipartMessage? = null
    
    @set:JvmSynthetic
    @get:JvmSynthetic
    internal lateinit var action: Action
    
    /**
     * 等待某个会话的回复，超时抛出超时异常
     * @param timeout 超时时间，单位为秒
     * @return  回复
     */
    @Throws(TimeoutException::class)
    fun waitForReply(timeout: Int): RobotMultipartMessage {
        //等待回复前，先忽略已有的回复
        reply = null
        var i = 0
        while(reply == null) {
            TimeUnit.SECONDS.sleep(1)
            i++
            if(i >= timeout) throw TimeoutException()
        }
        return reply!!
    }
    
    inline fun waitForReply(
        prompt: String,
        promptOnInvalidValue: String = "提供的参数有误，请重新输入",
        resultPredicate: (String) -> Boolean = { true },
        timeout: Int = 60
    ): String {
        reply(prompt)
        var result: String
        while(true) {
            try {
                result = waitForReply(timeout).contentToString()
                val isValid = runCatching { resultPredicate(result) }.getOrDefault(false)
                if(isValid) break
                reply(promptOnInvalidValue)
            } catch(t: Throwable) {
                if(t is TimeoutException) throw t
                reply(promptOnInvalidValue)
            }
        }
        return result
    }
    
    fun reply(message: RobotMultipartMessage) {
        sessionManager.framework.reply(group, qq, message)
    }
    
    fun reply(message: String) {
        sessionManager.framework.reply(group, qq, message)
    }
    
    @JvmSynthetic
    internal fun run() {
        try {
            action.action!!(this)
        } catch(t: Throwable) {
            if(t !is TimeoutException) throw t
            if(action.onTimeout != null) {
                action.onTimeout!!(this)
            } else {
                sessionManager.framework.reply(group, qq, "会话已超时关闭")
            }
        } finally {
            close()
        }
    }
    
    override fun close() {
        sessionManager.closeSession(this)
    }
}
